Report error in status if more than one DataScienceCluster installed

ykaliuta opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
Report error in status if more than one DataScienceCluster installed

V2
- update `phase` to `phaseError` as well
- add reporting for DSCInitialization

## Description
<!--- Describe your changes in detail -->

## How Has This Been Tested?

- Installed 2 DSCs in the same namespace with different name
- Checked `oc describe datasciencecluster.datasciencecluster.opendatahub.io/<NAME>` for both clusters

Both clusters report like 

```
Status:
  Conditions:
    Last Heartbeat Time:   2023-11-01T08:33:42Z
    Last Transition Time:  2023-11-01T08:32:20Z
    Message:               only one instance of DataScienceCluster object is allowed. Update existing instance on namespace  and name ykaliuta.dsc2
    Reason:                DuplicateDataScienceCluster
    Status:                False
    Type:                  ReconcileComplete
    Last Heartbeat Time:   2023-11-01T08:33:42Z
    Last Transition Time:  2023-11-01T08:32:20Z
    Message:               only one instance of DataScienceCluster object is allowed. Update existing instance on namespace  and name ykaliuta.dsc2
    Reason:                DuplicateDataScienceCluster
    Status:                False
    Type:                  Available
    Last Heartbeat Time:   2023-11-01T08:33:42Z
    Last Transition Time:  2023-11-01T08:31:39Z
    Message:               only one instance of DataScienceCluster object is allowed. Update existing instance on namespace  and name ykaliuta.dsc2
    Reason:                DuplicateDataScienceCluster
    Status:                False
    Type:                  Progressing
    Last Heartbeat Time:   2023-11-01T08:33:42Z
    Last Transition Time:  2023-11-01T08:32:20Z
    Message:               only one instance of DataScienceCluster object is allowed. Update existing instance on namespace  and name ykaliuta.dsc2
    Reason:                DuplicateDataScienceCluster
    Status:                True
    Type:                  Degraded
    Last Heartbeat Time:   2023-11-01T08:33:42Z
    Last Transition Time:  2023-11-01T08:32:20Z
    Message:               only one instance of DataScienceCluster object is allowed. Update existing instance on namespace  and name ykaliuta.dsc2
    Reason:                DuplicateDataScienceCluster
    Status:                Unknown
    Type:                  Upgradeable
    Last Heartbeat Time:   2023-11-01T08:32:12Z
    Last Transition Time:  2023-11-01T08:32:12Z
    Message:               Component reconciled successfully
    Reason:                ReconcileCompleted
    Status:                True
    Type:                  workbenchesReady
```

For DSCInitialization checked both DataScinceCluster:

```

  status:
    conditions:
    - lastHeartbeatTime: "2023-11-01T20:26:35Z"
      lastTransitionTime: "2023-11-01T20:26:14Z"
      message: only one instance of DSCInitialization object is allowed
      reason: DuplicateDSCInitialization
      status: "False"
      type: ReconcileComplete
    - lastHeartbeatTime: "2023-11-01T20:26:35Z"
      lastTransitionTime: "2023-11-01T20:26:14Z"
      message: only one instance of DSCInitialization object is allowed
      reason: DuplicateDSCInitialization
      status: "False"
      type: Available
    - lastHeartbeatTime: "2023-11-01T20:26:35Z"
      lastTransitionTime: "2023-11-01T20:26:14Z"
      message: only one instance of DSCInitialization object is allowed
      reason: DuplicateDSCInitialization
      status: "False"
      type: Progressing
    - lastHeartbeatTime: "2023-11-01T20:26:35Z"
      lastTransitionTime: "2023-11-01T20:26:14Z"
      message: only one instance of DSCInitialization object is allowed
      reason: DuplicateDSCInitialization
      status: "True"
      type: Degraded
    - lastHeartbeatTime: "2023-11-01T20:26:35Z"
      lastTransitionTime: "2023-11-01T20:26:14Z"
      message: only one instance of DSCInitialization object is allowed
      reason: DuplicateDSCInitialization
      status: Unknown
      type: Upgradeable
    phase: Error
```

and DSCIinitialization:

```status:
  conditions:
  - lastHeartbeatTime: "2023-11-01T20:25:08Z"
    lastTransitionTime: "2023-11-01T20:23:46Z"
    message: only one instance of DSCInitialization object is allowed. Update existing
      instance name new
    reason: DuplicateDSCInitialization
    status: "False"
    type: ReconcileComplete
  - lastHeartbeatTime: "2023-11-01T20:25:08Z"
    lastTransitionTime: "2023-11-01T20:23:46Z"
    message: only one instance of DSCInitialization object is allowed. Update existing
      instance name new
    reason: DuplicateDSCInitialization
    status: "False"
    type: Available
  - lastHeartbeatTime: "2023-11-01T20:25:08Z"
    lastTransitionTime: "2023-11-01T20:19:40Z"
    message: only one instance of DSCInitialization object is allowed. Update existing
      instance name new
    reason: DuplicateDSCInitialization
    status: "False"
    type: Progressing
  - lastHeartbeatTime: "2023-11-01T20:25:08Z"
    lastTransitionTime: "2023-11-01T20:23:46Z"
    message: only one instance of DSCInitialization object is allowed. Update existing
      instance name new
    reason: DuplicateDSCInitialization
    status: "True"
    type: Degraded
  - lastHeartbeatTime: "2023-11-01T20:25:08Z"
    lastTransitionTime: "2023-11-01T20:23:46Z"
    message: only one instance of DSCInitialization object is allowed. Update existing
      instance name new
    reason: DuplicateDSCInitialization
    status: Unknown
    type: Upgradeable
  phase: Error
```


<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Merge criteria:
<!--- This PR will be merged by any repository approver when it meets all the points in the checklist -->
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

ykaliuta commented:
Actually, should be revisited after issue #645 integration.

zdtsw commented:
ref: https://github.com/opendatahub-io/opendatahub-operator/issues/644

zdtsw commented:
do we want to add the same status update for DSCI as well?

ykaliuta commented:
> do we want to add the same status update for DSCI as well?

Good question. Probably it will not do any harm. Just IIUC it is more complicated for the user to encounter. Like install the operator twice. When deploying twice just the DataScienceCluster is much easier.

Well, if there are no  objections I'll add it as well. To both DSCI and DSC `Reconcile`.

zdtsw commented:
> > do we want to add the same status update for DSCI as well?
> 
> Good question. Probably it will not do any harm. Just IIUC it is more complicated for the user to encounter. Like install the operator twice. When deploying twice just the DataScienceCluster is much easier.
> 
> Well, if there are no objections I'll add it as well. To both DSCI and DSC `Reconcile`.

ok, lets do like this way. 
we can get this out first and create a new issue for the DSCI, then have another PR for that issue, how you feel?

ykaliuta commented:
> > > do we want to add the same status update for DSCI as well?
> > 
> > 
> > Good question. Probably it will not do any harm. Just IIUC it is more complicated for the user to encounter. Like install the operator twice. When deploying twice just the DataScienceCluster is much easier.
> > Well, if there are no objections I'll add it as well. To both DSCI and DSC `Reconcile`.
> 
> ok, lets do like this way. we can get this out first and create a new issue for the DSCI, then have another PR for that issue, how you feel?

I'm ok with several patches in one PR.

Just thought a bit about that. May be there should exist some sort of Webhook to prevent such configurations at the first place?

zdtsw commented:
> > > > do we want to add the same status update for DSCI as well?
> > > 
> > > 
> > > Good question. Probably it will not do any harm. Just IIUC it is more complicated for the user to encounter. Like install the operator twice. When deploying twice just the DataScienceCluster is much easier.
> > > Well, if there are no objections I'll add it as well. To both DSCI and DSC `Reconcile`.
> > 
> > 
> > ok, lets do like this way. we can get this out first and create a new issue for the DSCI, then have another PR for that issue, how you feel?
> 
> I'm ok with several patches in one PR.
> 
> Just thought a bit about that. May be there should exist some sort of Webhook to prevent such configurations at the first place?

i do not want to enforce git webhook (if this is the webhook you meant) or it will be a new commit everytime and difficult for reviewer, we do have squash-merge for PR so the committer (or reviewer) is responsibile to get a clean history for the PR before merge

ykaliuta commented:
> > > > > do we want to add the same status update for DSCI as well?
> > > > 
> > > > 
> > > > Good question. Probably it will not do any harm. Just IIUC it is more complicated for the user to encounter. Like install the operator twice. When deploying twice just the DataScienceCluster is much easier.
> > > > Well, if there are no objections I'll add it as well. To both DSCI and DSC `Reconcile`.
> > > 
> > > 
> > > ok, lets do like this way. we can get this out first and create a new issue for the DSCI, then have another PR for that issue, how you feel?
> > 
> > 
> > I'm ok with several patches in one PR.
> > Just thought a bit about that. May be there should exist some sort of Webhook to prevent such configurations at the first place?
> 
> i do not want to enforce git webhook (if this is the webhook you meant) or it will be a new commit everytime and difficult for reviewer, we do have squash-merge for PR so the committer (or reviewer) is responsibile to get a clean history for the PR before merge

No, I meant kubernetes validating webhook, like https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation

zdtsw commented:
> > > > > > do we want to add the same status update for DSCI as well?
> > > > > 
> > > > > 
> > > > > Good question. Probably it will not do any harm. Just IIUC it is more complicated for the user to encounter. Like install the operator twice. When deploying twice just the DataScienceCluster is much easier.
> > > > > Well, if there are no objections I'll add it as well. To both DSCI and DSC `Reconcile`.
> > > > 
> > > > 
> > > > ok, lets do like this way. we can get this out first and create a new issue for the DSCI, then have another PR for that issue, how you feel?
> > > 
> > > 
> > > I'm ok with several patches in one PR.
> > > Just thought a bit about that. May be there should exist some sort of Webhook to prevent such configurations at the first place?
> > 
> > 
> > i do not want to enforce git webhook (if this is the webhook you meant) or it will be a new commit everytime and difficult for reviewer, we do have squash-merge for PR so the committer (or reviewer) is responsibile to get a clean history for the PR before merge
> 
> No, I meant kubernetes validating webhook, like https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation

ok, but then it is a totally different story which will add new webhook in api to validate fields etc.
I started this way  a couple of months ago but did not continue with it due to other high priority tasks, but if you would like you can create a ticket and work on this.

ykaliuta commented:
> > > > Just thought a bit about that. May be there should exist some sort of Webhook to prevent such configurations at the first place?
> > > 
> > > 
> > > i do not want to enforce git webhook (if this is the webhook you meant) or it will be a new commit everytime and difficult for reviewer, we do have squash-merge for PR so the committer (or reviewer) is responsibile to get a clean history for the PR before merge
> > 
> > 
> > No, I meant kubernetes validating webhook, like https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation
> 
> ok, but then it is a totally different story which will add new webhook in api to validate fields etc.

Sure, it's a further enhancement and it's good to get proper reporting first (I've updated the PR BTW and added phaseError set like in some other places)

> I started this way a couple of months ago but did not continue with it due to other high priority tasks, but if you would like you can create a ticket and work on this.

Sounds great actually!

zdtsw commented:
create issue : https://github.com/opendatahub-io/opendatahub-operator/issues/693

zdtsw reviewed (approved):
/lgtm

zdtsw commented on controllers/datasciencecluster/datasciencecluster_controller.go:
```diff
@@ -137,7 +142,12 @@ func (r *DataScienceClusterReconciler) Reconcile(ctx context.Context, req ctrl.R
 		dscInitializationSpec.DeepCopyInto(r.DataScienceCluster.DSCISpec)
 	default:
-		return ctrl.Result{}, errors.New("only one instance of DSCInitialization object is allowed")
+		message := "only one instance of DSCInitialization object is allowed"
```
why do we have check again on the number of DSCI instance? 
if we have it here, do we still need the block if len(instances.Items) > 1 {...}

ykaliuta replied:
In datasciencecluster_controller.go file? `if len(instances.Items) > 1 `  checks for DataScienceCluster objects, but here it is for DSCInitialization one.
